packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -1157,12 +1169,31 @@ export class DeltaManager
                 if (duplicateEnd === undefined || duplicateEnd < message.sequenceNumber) {
                     duplicateEnd = message.sequenceNumber;
                 }
+
+                // Validate that we do not have data loss, i.e. sequencing is reset and started again
+                // with numbers that this client already observed before.
+                if (this.previouslyProcessedMessage?.sequenceNumber === message.sequenceNumber) {
+                    const message1 = this.comparableMessagePayload(this.previouslyProcessedMessage);
+                    const message2 = this.comparableMessagePayload(message);
+                    if (message1 !== message2) {
+                        const error = {
+                            errorType: ContainerErrorType.dataCorruption,
+                            message: "Two messages with same seq# and different payload!",
+                            clientId: this.connection?.details.clientId,
+                            sequenceNumber: message.sequenceNumber,
+                            message1,
+                            message2,
+                        };
+                        this.close(error);
+                    }
+                }
```

Comment:
I am not sure what we are checking here. This code already already eliminates duplicates. So why are we throwing error here?

Are we only checking service level invariance here? If the answer is yes, then this only covers a very specific scenario. Referring back to my earlier comment, we are only checking whether op# 53 is altered. We still cannot check 50, 51, and 52 and have no idea what happened to them.

Rather than checking here, how about we check our protocol level invariance? Such as why msn moves backward and log the incoming op that probably has an incorrect reference sequence number. Or a "leave" without a "join", log the "leave" message. I am sure there are more places where we "assert" today. Service can check those logs and figure out the root cause.

Reply:
You are right that this checks only one op. That said, if storage screwed up somewhere (without telling client about change in epoch/term) and lost ops, chances that we will pass that check with new ops is close to zero. From that POV, we do not care that much about prior ops, as client (on reconnect) needs to see its own join op before sending any new ops, so it will fail (with this check) before it corrupts document.

You are right that there are other less direct checks in code like msn moving backward or quorum that are likely to catch this issue. That said, this check is much stronger, and much sooner, and more direct, i.e. developers will not scratch their heads looking at unrelated asserts here and there. It's both defense in depth and grouping all such cases into one bucket in telemetry, allowing us much easier quantify and notice the problem if/when it starts happening.

Reply:
Just to be clear, we cannot assume that duplication will only happen after disconnect. That was true in old Push world. In push V2, sequencer (deli) is separated from socket box (alfred). And we don't disconnect clients on "deli" restart cause the client code handles the duplication. So even with this barrier, those asserts will still trigger and clients will keep sending ops until you reach the last op.

Reply:
I have hard time following. First, there is nothing about disconnect in this code.
True duplication (i.e. exactly same content) is fine - there is no problem with that, it will pass new check just fine and we will dedup ops as before.
duplication as sequence reuse with different content is not Ok, it does not matter how that happens, it's a bug that we want to catch and fail.

> So even with this barrier, those asserts will still trigger and clients will keep sending ops until you reach the last op.

Not sure I follow. Most asserts on op processing are exceptions that close container - no ops would be sent after that. The code I've added even stronger - it does not rely on someone catching exception and doing right thing - it closes container permanently.

Reply:
> From that POV, we do not care that much about prior ops, as client (on reconnect) needs to see its own join op before sending any new ops, so it will fail (with this check) before it corrupts document.

I thought you are assuming disconnection as a prerequisite of this error. Just wanted to clarify that this is not the case. The bug you are tackling can happen within a connection.

The op you are checking (op 53) will be discarded anyway. So even if the content is altered here, you don't care much. I think the implicit assumption is, if op 53 is busted, op 54 and rest will be busted as well. So fail sooner rather than failing on asserts later. This is a valid assumption (with some caveat) but it still does not fix the issue of #3840. For example: Leave without Join can happen just because service failed to sequence the join op. msn moving backward can happen cause we have a service bug in msn calculation even if the contents are fine. So we still need checks in specific places even if this check is present.

Reply:
Sure, I'm not removing any existing checks

Reply:
So today, if we fail on an assert, do we log those in telemetry? More specifically, do we log the "op" that caused the assert to trigger?

Reply:
We never log ops, as they contain PII. We do log sequence number for any error that results in container being closed:
```
            this.logger.sendErrorEvent(
                {
                    eventName: "ContainerClose",
                    sequenceNumber: error.sequenceNumber ?? this._deltaManager.lastSequenceNumber,
                    loading: !this.loaded,
                },
                error,
            );
```
that said, usefulness of this data greatly depends on access to a file.

As for asserts, it really depends on where it is thrown and how is catching it. For op processing, pretty much everything goes to DeltaQueue that catches exceptions and raises "error" event that goes to DeltaManager -> container close.
Likely can be improved (move this logic out of DeltaQueue to be directly in DeltaManager's handlers.

Reply:
Without knowing details about the assert flow, my expectation is that we should log the "op" that caused the assert. That will be incredibly valuable for the server side in case it missed to store the op completely. Regarding PII, I also have very little knowledge. But we can probably just delete the content field and upload the rest?

Reply:
For the "msn moves backwards" case, I agree with tanvir that we should add at least the sequenceNumber and msn of the op that caused the assert.
@vladsud the sequenceNumber in the log that you pointed to will give us the sequenceNumber of the op that delta manager processed before the op that caused the assert. I think we should log both op's sequenceNumber.

Reply:
Let's not pile everything into same PR :)
If we need to improve existing asserts, let's do that in separate PRs. But we already log currently processed sequence number when closing container (for all errors), so it's a matter to use it directly, or substract one (depending on where assert happened - before we did ++ on variable or after).
Same of the op - we can do it (in same close() code) and collect more data. That said, I do not think it would be super useful - we already have all the interesting data to do most of the analyzes or correlation to server logs (including clientID, sequenece numbers, etc.)

Reply:
Ditto on separate PRs. On the correlation part, I agree with you but the service may not always have a log to check against. If logging the whole op (minus the content) seems wasteful, we can discuss what we want to log for what type of op as well. Example: For system ops such as join and leave, we should log the "data" field.

Reply:
Privacy violation!: ) data filed contains too much info about user.
We can't log emails or any PII. I believe (would need to check if that's still the case), Aria would drop whole event if it can detect there is PII like email - we usually notice lack of data simply by accident, as I'm not aware of direct way of getting this data.
Basically, we should only log that we are 100% sure does not contain PII.

Let's move this discussion somewhere else :)